modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java

```
@@ -364,16 +387,34 @@ private V compute(ICacheKey<K> key, LoadAwareCacheLoader<ICacheKey<K>, V> loader
                     future.completeExceptionally(npe);
                     throw new ExecutionException(npe);
                 } else {
-                    future.complete(new Tuple<>(key, value));
+                    if (evaluatePoliciesList(value, policies)) {
+                        future.complete(new Tuple<>(key, value));
+                        wasCacheMiss = true;
+                    } else {
+                        future.complete(null); // Passing null would skip the logic to put this into onHeap cache.
+                        // Signal to the caller that the key didn't enter the cache by sending a removal notification.
+                        // This case does not count as a cache miss.
+                        removalListener.onRemoval(new RemovalNotification<>(key, value, RemovalReason.EXPLICIT));
+                        wasRejectedByPolicy = true;
+                    }
                 }
             } else {
                 try {
-                    value = future.get().v2();
+                    Tuple<ICacheKey<K>, V> futureTuple = future.get();
+                    if (futureTuple == null) {
+                        // This case can happen if we earlier completed the future with null to skip putting the value into the cache.
+                        // It does not count as a cache miss.
+                        value = loader.load(key);
+                        removalListener.onRemoval(new RemovalNotification<>(key, value, RemovalReason.EXPLICIT));
+                        wasRejectedByPolicy = true;
```

Looking at this again, doesn't look right. Seems like we can end up having the loading the value twice in case caching policies are not satisfied? This would degrade the performance!

---

I think one way to handle this would be to maybe change the future handler to accept a `boolean` which basically sends wasRejectedByPolicy as true/false. And then handler can use this value to accordingly skip adding item onto heap cache.

---

In addition to this, I don't we should be calling `removalListener.onRemoval(new RemovalNotification<>(key, value, RemovalReason.EXPLICIT));` here as well?  
As this else condition is triggered when a different thread T1 is already loading the value for same key, and the current thread(T2) just waits on that for value to be loaded. In this case IRC would  record a hit for request on T2 and not a miss?

---

You're right, I'll change it to prevent double-loading. 

If we change it so that T2 can return the value loaded by T1, even if the value never enters the cache because it was rejected by the policy, should that count as a cache hit? I'm not sure either way. I think it would probably make more sense for it to not count as either a hit or a miss, since we aren't counting the case where there's a single thread with a rejected value as a miss either. 

I do think either way we should still have the removal notification on line 408. We decided it made sense to have it on line 397 to alert the outside caller that their value never entered the cache, and this case seems the same to me.

---

Adding a UT for this is tricky. I'm not sure of a way to make sure the second rejected key gets to `future.get()` before the first rejected key has finished running `computeIfAbsent` and had its future removed from the map. If it finishes later, it will load the key again which is expected. The only thing we can inject is logic in the loader, but the loader must completely finish before the first key can run `future.complete()`, so we can't add any latch in there.

---

>If we change it so that T2 can return the value loaded by T1, even if the value never enters the cache because it was rejected by the policy, should that count as a cache hit?

Yeah, we shouldn't count this as a hit or a miss like you mentioned. But I guess at IRC level, it has to be either of hit/miss and we can't say it wasn't neither, right? 

Regarding removal notification, my concern was that the old stats would break. As for request on T2 which is waiting on same key being loaded on T1, removal notification on T2 would cause old stats to deduct duplicate (key + value) size  and potentially go negative. 
I am questioning my own previous thought here - whether we should really invoke removal notification in case policy isn't satisfied? Maybe no. Then we assume that TC just silently ignore this key/value for now, and we can rethink this in future if its really needed. What do you think? Would not calling removal notification cause any issues elsewhere?

---

I think at this point the old IRC stats are unfixable if we want to go forward with having the took time policy in front of the heap tier for TSC. There's too many cases dependent on logic within the TSC that the IRC can't/shouldn't know about. But like we discussed we're getting rid of these stats in 3.0 so it should be fine. Getting rid of the removal notification shouldn't cause issues anywhere except the old IRC stats, so I can do that.

---

Thanks for the change. Looks good to me now!